[ASTGen] Introduce wrapper types for AST nodes

hamishknight opened the pull request:
Stamp out wrapper types for all the AST nodes, and use them for ASTGen, with members being imported on those types.

Resolves #68346
Resolves #68348
rdar://117159010

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@ahoppen For #68348 my preference is to import as static members since this would make it easier to switch to C++ interop

ahoppen commented on include/swift/AST/CASTBridging.h:
```diff
@@ -31,6 +31,69 @@
 #endif
 
 SWIFT_BEGIN_NULLABILITY_ANNOTATIONS
+
+typedef struct BridgedDeclContext {
+  void *_Nonnull raw;
+} BridgedDeclContext;
```
Why can’t we use the `AST_BRIDGING_WRAPPER_NONNULL` macro for `BridgedDeclContext`?

hamishknight replied:
Good catch, this was needed in an earlier revision to avoid defining `bridged(DeclContext *)`, but I don't think it actually causes any issues any more, I'll change it

ahoppen commented on include/swift/AST/CASTBridging.h:
```diff
@@ -31,6 +31,69 @@
+extern "C" {
+#endif
+
+// Declare `.asDecl` on each BridgedXXXDecl type.
```
Conceptually these are upcasts, right? Might be worth pointing out in the comment.

ahoppen commented on lib/AST/CASTBridging.cpp:
```diff
@@ -44,58 +44,116 @@ struct BridgedDiagnosticImpl {
+// Define `.asDecl` on each BridgedXXXDecl type.
+#define DECL(Id, Parent)                                                       \
+  BridgedDecl Id##Decl_asDecl(Bridged##Id##Decl decl) {                        \
+    return bridged((Decl *)unbridged(decl));                                   \
```
Any opinion on whether to use `static_cast` here? I think (not sure) that it does more checking and should be sufficient here. I don’t have really strong opinions here, though. So up to you.

hamishknight replied:
Heh yeah, I didn't really care since we know up-front that it's always an upcast, but yeah you're right `static_cast` is better

ahoppen commented on lib/AST/CASTBridging.cpp:
```diff
@@ -44,58 +44,116 @@ struct BridgedDiagnosticImpl {
 };
 } // namespace
 
+static inline BridgedDeclContext bridgedDeclContext(DeclContext *declContext) {
+  return {declContext};
+}
+
+static inline DeclContext *unbridged(BridgedDeclContext cDeclContext) {
+  return static_cast<DeclContext *>(cDeclContext.raw);
+}
```
In Swift, these would make wonderful extensions on `BridgedDeclContext` and `DeclContext`. A shame we don’t have type extensions in C++. 😢

ahoppen commented on lib/AST/CASTBridging.cpp:
```diff
@@ -255,104 +314,113 @@ void *FunctionCallExpr_create(BridgedASTContext cContext, void *fn,
+                                        BridgedArrayRef commas,
+                                        BridgedSourceLoc cRLoc) {
+  ASTContext &context = unbridged(cContext);
+  return bridged(ArrayExpr::create(
+      context, unbridged(cLLoc), unbridgedArrayRef<Expr *>(elements),
+      unbridgedArrayRef<SourceLoc>(commas), unbridged(cRLoc)));
```
Not really important but do we have a somewhat consistent style if we do `return bridged(XXX::create)` or 
```cpp
auto X = XXX::create…`
return bridged(X);
```

I would have probably used the latter version because it means that the parameters aren’t nested as much. But it’s not all that important.

hamishknight replied:
I think for single-line cases, nesting it in the `bridged(...)` call is fine, but yeah maybe we should split out the multi-line cases

ahoppen commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -13,35 +13,47 @@ extension UnsafePointer {
     }
+    return bridged
+  }
+
+  var castToType: BridgedTypeRepr {
+    guard case .type(let bridged) = self else {
+      fatalError("Expected a type")
+    }
+    return bridged
   }
 
   var bridged: BridgedASTNode {
     switch self {
     case .expr(let e):
-      return BridgedASTNode(ptr: e, kind: .expr)
+      return BridgedASTNode(ptr: e.raw!, kind: .expr)
```
Should we create a `BridgedASTNode` with a `nullptr` instead of crashing here? AFAICT nothing guarantees that `raw` is not `nil` here.

hamishknight replied:
`BridgedASTNode` has a non-nullable value, so we have to unwrap here. We could change `BridgedASTNode` to take a nullable value, but that would just be trading an explicit unwrap for an implicit null pointer access in C++ land. The alternative here would be to introduce separate nullable types for the base AST node wrapper types, e.g `BridgedExpr` and `BridgedNullableExpr`, and then parameters expecting a nullable expr could use the nullable version and you'd have to add an extra `.init(...)` when passing the argument. I can't say I'm a huge fan of that, though I do agree the force unwrap here isn't great.

ahoppen replied:
Or a completely different option: Do you think we actually need `ASTNode` or could we remove it altogether (definitely not this PR). To me `ASTNode` seems like an anti-pattern anyway because it generalizes too much. 

For example `ASTGen.generate(ExprSyntax)` should switch over all the expression nodes (for which we need to implement an `ExprSyntaxEnum`, which I think I have thought about adding for 2 years now) and then it could just return a `BridgedExprSyntax` without the need to go through `ASTNode`.

hamishknight replied:
The issue is for things like BraceStmt, where the elements truly are ASTNodes :/

hamishknight replied:
FWIW this is what it would look like with nullable wrappers ef7407804c87f2bd3826ab57c131fe73f8da24fe, I guess it's not _too_ bad? Still can't say I love it

ahoppen replied:
I think I prefer it with https://github.com/apple/swift/commit/ef7407804c87f2bd3826ab57c131fe73f8da24fe. At least it’s explicit about what it’s doing and the `asNullable` are annoying but not toooo bad. But it’s Friday evening and I might change my mind when I look at this again on Monday.

hamishknight replied:
Yeah, I don't really have particulary strong feelings either way, I've left it as-is for now, but am happy to do the nullable wrappers in a follow up

hamishknight replied:
Hmm the more I think about, the less I hate it, put up a PR: #69326

ahoppen commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -314,18 +318,18 @@ extension ASTGenVisitor {
+  func generate(_ node: TypeSyntax?) -> BridgedTypeRepr {
     guard let node else {
-      return nil
+      return .init(raw: nil)
```
1. I’m not a huge fan of `.init` because it’s usually unclear which type is being created. `BridgedTypeRepr(raw: nil)` is a little bit longer to write but a lot easier to read IMO.
2. Have you considered conforming `BridgedTypeRepr` etc. to `ExpressibleByNilLiteral`?

hamishknight replied:
I wanted to keep this fairly verbose since ASTGen shouldn't need to construct `nil` AST nodes all that often, and it should be made clear that it's burying the `nil` value in a non-optional type. I'm happy to change to `BridgedTypeRepr(raw: nil)`

ahoppen replied:
Sounds good to me. Changing to `BridgedTypeRepr(raw: nil)` would be appreciated.

ahoppen reviewed (approved):
Soooo much nicer. Thank you!

A couple of ideas to improve things further inline but as far as I’m concerned nothing that’s blocking us from merging the PR.

hamishknight committed:
[AST] NFC: Introduce `ABSTRACT_CONTEXT_DECL` macro

This is for decl nodes that are both abstract and
are DeclContexts.

hamishknight committed:
[ASTGen] Introduce wrapper types for AST nodes

Stamp out wrapper types for all the AST nodes,
and use them for ASTGen, with members being
imported on those types.

hamishknight committed:
[ASTGen] Use `SWIFT_NAME` in some more places

hamishknight committed:
Address review feedback

hamishknight committed:
Rename `convertXXX` functions to `unbridged`

This is consistent with the AST node bridging
naming scheme and IMO the old names were
needlessly verbose.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test macOS

hamishknight commented:
@swift-ci please test macOS
